### DISCUSSION TITLE
Anyone have a sample Python client to do a simple query using GraphQL?

### DISCUSSION THREAD

**Discussion by kurtis318** [requarks/wiki#3431] [Help / Questions]
**Question**
I'm new to Wiki.js and GraphQL.  I have studied the 4 videos at https://www.howtographql.com/ and they make sense to me.  I was expecting the Wiki.js documentation at https://docs.requarks.io/dev/api to have some sample code.  Does anyone know of any publicly available sample code in Python or any other language?  I know the GraphQL schema should describe the api I just don't know how the authentication works.  Thanks.

**Host Info (please complete the following information):**
OS: Fedora 32 Server (native)
Wiki.js version: 2.4.107
Database engine: postgres 12.3-1

**Comment by theriverman** [comment]
I feel you. The GraphQL API documentation is quite poor at the moment, and I've had a hard time too to get my head around it.
I'm pretty-pretty far from being a GraphQL expert, but maybe I can get your started.

Your best references are the [source code ](https://github.com/Requarks/wiki/blob/dev/server/graph/schemas/authentication.graphql), and the `Network` tab in your developer console (press `F12`).
Do something in your Wiki.js instance, and see what gets posted to the `/graphql` endpoint. Using the sent payload, you can figure our how should your custom query look like.

To authenticate, that is, to get a JWT, your query should look like this:
```graphql
mutation ($username: String!, $password: String!, $strategy: String!) {
    authentication {
        login(username: $username, password: $password, strategy: $strategy) {
            responseResult {
                succeeded
                errorCode
                slug
                message
                __typename
            }
            jwt
            mustChangePwd
            mustProvideTFA
            continuationToken
            __typename
        }
        __typename
    }
}
```

and obviously the additional variables should be:
```json
{
    "username": "your-username",
    "password": "your-password",
    "strategy": "local"
}
```
( Hint: If you're using AD/LDAP authentication, the strategy should be set accordingly. I leave it to you to figure that out :) )

Here's a Python code example. It's very basic, it does not make use of GraphQL schemas at all, but it should be sufficient to get you started:
(Used library: https://github.com/profusion/sgqlc )
```python
from sgqlc.endpoint.http import HTTPEndpoint

url = "https://your-wiki-server.company/graphql"
# Add your JWT here once you have it. Leave dict empty for authentication
headers = {'Authorization': 'Bearer <your-very-long-jwt-comes-here>'}

q = """query ($userID: Int!) {
    users {
        single(id: $userID) {
            id
            name
            email
            providerKey
            providerId
            location
            jobTitle
            timezone
            isSystem
            isActive
            isVerified
            createdAt
            updatedAt
            lastLoginAt
            groups {
                id
                name
            }
        }
    }
}"""
endpoint = HTTPEndpoint(url, headers)
results = endpoint(q, {'userID': 1})
print(results)

```

**Comment by kurtis318** [comment]
@theriverman Thank you for the example and taking the time.  I really appreciate your post.

**Comment by zhuzhzh** [comment]
I am trying to use `tree`, but it doesn't work.

```txt
from sgqlc.endpoint.http import HTTPEndpoint

url = "http://wiki.etmcu.com/graphql"

headers = {'Authorization': 'Bearer  <api key>'}

q = """query {
  pages {
    tree (mode: PAGES, locale: "en") {
      id
      path
      pageId
    }
  }
}"""
endpoint = HTTPEndpoint(url, headers)
results = endpoint(q, {})
print(results)
```

The error message is as below:

```txt
GraphQL query failed with 1 errors
{'errors': [{'message': 'builder.andWhereNotNull is not a function', 'locations': [{'line': 3, 'column': 5}], 'path': ['pages', 'tree'], 'extensions': {'code': 'INTERNAL_SERVER_ERROR'}}], 'data': {'pages': {'tree': None}}}
```

Is there anything wrong above?
